pubsub/pubsub.go

```diff
@@ -15,22 +15,36 @@ type Message interface{}
 // Publish publishes a message to a channel.
 func Publish(channel string, data Message) {
 	mu.Lock()
+	defer mu.Unlock()
 	for _, listener := range listeners[channel] {
-		if listener != nil {
-			listener <- data
-		}
+		listener <- data
 	}
-	mu.Unlock()
 }
 
 // Subscribe subscribes to the channel and returns listener for it.
 func Subscribe(channel string) chan Message {
-	listener := make(chan Message)
 	mu.Lock()
+	defer mu.Unlock()
+	listener := make(chan Message)
 	if listeners[channel] == nil {
 		listeners[channel] = make([]chan Message, 0)
 	}
 	listeners[channel] = append(listeners[channel], listener)
-	mu.Unlock()
 	return listener
 }
+
+// Unsubscribe unsubscribes a listener from listening channel.
+func Unsubscribe(channel string, listener chan Message) {
+	mu.Lock()
+	defer mu.Unlock()
+	if len(listeners[channel]) == 1 {
```

Comment:
I'm not sure about that one. What happen if for any reason (by design it shouldn't be possible but everything is possible) the list of `listeners[channel]` contains another `chan` than `listener`.

I would prefer to iterate over the `listeners` and if it's just about memory optimization we can still at the end of the loop, if the length of `listeners[channel]` is 0 then delete it

Is there any other reason why you did this condition here ?

Reply:
just a memory optimization. if it contains more than 1 listener then program is going to skip the if statement and **a** listener will be removed. This will happen until there is only 1 one listener left in the slice and for that one execution will go through the if statement and slice will be nilled.

I think both ways are totally of equivalent each other. If you prefer the second way as a pattern, I don't mind to change.

Reply:
I'm just afraid about some possible side effects like another part add a new chan in the `listeners[channel]` directly and then there is an `unsubscribe` called with nil for the listener for example. In this case the first chan added will be deleted even we unsubscribed with nothing.

This can of thing should not happen but if for any reasons it happen we will have a side effect here but if we clear the memory when the slice is empty already (after the check of the chan) then it's no problem

Reply:
I don't understand: `I'm just afraid about some possible side effects like another part add a new chan in the listeners[channel] directly`

A channel to `listeners[channel]` only added within `Subscribe()`

Reply:
Call to a `Unsubscribe()` will not remove any channel from `listeners[channel]` because we compare pointer values of channels.

Reply:
oh, i got it. i'll change the line

Reply:
done